### PR TITLE
Disable use of default credentials in windows with Tor.

### PR DESCRIPTION
**Pull request by riastradh-brave**
fix brave/brave-browser#3603

- New class AnonHttpAuthPreferences is like HttpAuthPreferences, but
  unconditionally disables the use of default credentials.

- Each TorProfileServiceImpl owns an AnonHttpAuthPreferences object.

- OnBeforeURLRequest_TorWork sets the auth preferences to be the Tor
  profile's anonymous auth preferences.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
https://hackerone.com/reports/505764

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

File: browser/net/anon_http_auth_preferences.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2018 The Brave Authors. All rights reserved.
```
Comment (simonhong) [inline browser/net/anon_http_auth_preferences.h]: nit: 2019 not 2018.
And all other files we created uses below.
```
/* Copyright (c) 2019 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */
```

Comment (riastradh-brave) [reply browser/net/anon_http_auth_preferences.h]: Whoops, copypasta from another file; lint didn't catch it.  Fix pushed.

**Comment by bbondy** [comment]
This PR is getting pretty old, just checking in on if it's in deadlock?

**Comment by riastradh-brave** [comment]
I haven't been able to conclusively determine that the issue is fixed.  The browser still offers NTLM authentication even with this patch.  However, I realized after various experiments to confirm that this logic is being used that it's _possible_ that the NTLM authentication doesn't actually leak default credentials -- that although the browser initiates NTLM authentication, it's not actually providing any credentials.

**Comment by abartlet** [comment]
If it starts NTLM it might still be leaking a hostname:

from Samba's ntlmssp.idl

```
    typedef [public] struct {
                [charset(DOS),value("NTLMSSP")] uint8 Signature[8];
                [value(NtLmNegotiate)] ntlmssp_MessageType MessageType;
                NEGOTIATE NegotiateFlags;
                [value(DomainName ? strlen(DomainName) : 0)] uint16 DomainNameLen;
                [value(DomainNameLen)] uint16 DomainNameMaxLen;
                [relative] [subcontext(0),subcontext_size(DomainNameLen)] [flag(ndr_ntlmssp_negotiated_string_flags(NTLMSSP_NEGOTIATE_OEM))] string *DomainName;
                [value(Workstation ? strlen(Workstation) : 0)] uint16 WorkstationLen;
                [value(WorkstationLen)] uint16 WorkstationMaxLen;
                [relative] [subcontext(0),subcontext_size(WorkstationLen)] [flag(ndr_ntlmssp_negotiated_string_flags(NTLMSSP_NEGOTIATE_OEM))] string *Workstation;
                [switch_is(NegotiateFlags & NTLMSSP_NEGOTIATE_VERSION)] ntlmssp_Version Version;
        } NEGOTIATE_MESSAGE;
```

I wouldn't want NTLM anywhere near a TOR browser.

**Comment by bbondy** [comment]
Is this still valid? No updates for around a month. Thanks.

**Comment by bsclifton** [comment]
Closing as stale. We can keep the branch of course, in case we wanted to revisit

**Pull request closed without merging**